### PR TITLE
JDK-8264222: Use switch expression in jshell where possible

### PR DESCRIPTION
**Pull request by amaembo**
This change is powered by IntelliJ IDEA quick-fix.
I also took the liberty to fix C-style arrays in changed files.

In SourceCodeAnalysisImpl.java:501, I left case PARAMETERIZED_TYPE -> FALSE as a separate switch case, because it had a comment. Better formatting ideas are welcome.

### CODE REVIEW

**Review by briangoetz — APPROVED** [review]
Cute trick for the partial update to `c`.  
The comment in the last hunk (resultTypeOf()) only applied to the INSTANCE and STATIC init cases, but now it applies to CONSTRUCTOR too (in a previous hunk, you deliberately had a redundant case, to preserve this information.)   I would update the comment.

Otherwise, all looks good to me, +1.

**Comment by amaembo** [comment]
@briangoetz thank you for review!

> I would update the comment.

I updated the comment and moved it before the case branch. Please take a look.

> Cute trick for the partial update to c.

Yeah, IDEA does this automatically!

**Review by briangoetz — APPROVED** [review]

**Comment by stuart-marks** [comment]
I have some formatting questions. I also have some weak opinions on formatting, but not strong enough to make recommendations yet. (I see you wrote "Better formatting ideas are welcome" so thanks for being open to this.) I'm interested in feedback on how the code ends up _reading_, particularly from the people who will be maintaining this code. But comments from others are welcome too.

The switch (c) in `ArgTokenizer::nextToken` is pretty nicely formatted, since each case has a single constant, they're all short, and the values are also all constants that are textually short. The only thing I'd say here is to add a bit of spacing to the default case so that the arrows line up.

In `Eval::prettyExpr` switch (typeName), the switch is a bit irregular since one of the case arms has multiple constants. This makes it a bit harder to find one if you're looking for it in particular. For example, suppose you want to find out how "int" is handled. In the old code you can scan down looking at each `case` and then look at the constant right next to it. In the new code (as modified) the "int" is farther away from the `case` keyword and is a bit lost among the values, so it's harder to see. Let me see how the variations look.

Here's the modified version as proposed in the PR:

```
    String sinit = switch (typeName) {
        case "byte", "short", "int" -> "0";
        case "long" -> "0L";
        case "float" -> "0.0f";
        case "double" -> "0.0d";
        case "boolean" -> "false";
        case "char" -> "'\\u0000'";
        default -> "null";
    };
```

Variation with the arrows on the same line, but lined up:

```
    String sinit = switch (typeName) {
        case "byte", "short", "int" -> "0";
        case "long"                 -> "0L";
        case "float"                -> "0.0f";
        case "double"               -> "0.0d";
        case "boolean"              -> "false";
        case "char"                 -> "'\\u0000'";
        default                     -> "null";
    };
```

Variation with arrows on the next line:

```
    String sinit = switch (typeName) {
        case "byte", "short", "int"
            -> "0";
        case "long"
            -> "0L";
        case "float"
            -> "0.0f";
        case "double"
            -> "0.0d";
        case "boolean"
            -> "false";
        case "char"
            -> "'\\u0000'";
        default
            -> "null";
    };
```

Variation with arrows on the same line, aligned, with case constants one per line:

```
    String sinit = switch (typeName) {
        case "byte",
             "short",
             "int"     -> "0";
        case "long"    -> "0L";
        case "float"   -> "0.0f";
        case "double"  -> "0.0d";
        case "boolean" -> "false";
        case "char"    -> "'\\u0000'";
        default        -> "null";
    };
```

**Comment by briangoetz** [comment]
I'll answer from two perspectives: what I personally like, vs what I 
would recommend in a style guide.

 From a personal perspective, I like #2 the best, and #4 after that.  It 
is really easy to see what's going on.  But, cases where you can 
actually line up like this are rare, since most switches involve more 
than just simple constants on the RHS.

On the other hand, I would never ask anyone to code like that; it seems 
too obsessive to ask people to do.  What I'd suggest from a style guide 
perspective is #1.

For this particular example, #3 is the worst on both points.

But, this particular example is specific; both the predicates and 
consequences are syntactically short.  If they were longer, #3 would be 
more attractive.

So as a general recommendation #1 if it fits, #3 if #1 doens't fit, and 
#2/#4 if the formatting works out and you're feeling generous.

**Comment by stuart-marks** [comment]
Right, with longer case constants or values, the formatting dynamics of the entire expression change. Here's another example from the changeset, from `Eval::kindOfTree`.

(5) Here's the version as proposed in the PR:

```
    OuterWrap outer = switch (probableKind) {
        case IMPORT -> state.outerMap.wrapImport(Wrap.simpleWrap(compileSource), snip);
        case EXPRESSION -> state.outerMap.wrapInTrialClass(Wrap.methodReturnWrap(compileSource));
        case VAR, TYPE_DECL, METHOD -> state.outerMap.wrapInTrialClass(Wrap.classMemberWrap(compileSource));
        default -> state.outerMap.wrapInTrialClass(Wrap.methodWrap(compileSource));
    };
```

(6) Line breaks before the arrow:

```
    OuterWrap outer = switch (probableKind) {
        case IMPORT
            -> state.outerMap.wrapImport(Wrap.simpleWrap(compileSource), snip);
        case EXPRESSION
            -> state.outerMap.wrapInTrialClass(Wrap.methodReturnWrap(compileSource));
        case VAR, TYPE_DECL, METHOD 
            -> state.outerMap.wrapInTrialClass(Wrap.classMemberWrap(compileSource));
        default
            -> state.outerMap.wrapInTrialClass(Wrap.methodWrap(compileSource));
    };
```

(7) An extra blank line between each case:

```
    OuterWrap outer = switch (probableKind) {
        case IMPORT
            -> state.outerMap.wrapImport(Wrap.simpleWrap(compileSource), snip);

        case EXPRESSION
            -> state.outerMap.wrapInTrialClass(Wrap.methodReturnWrap(compileSource));

        case VAR, TYPE_DECL, METHOD 
            -> state.outerMap.wrapInTrialClass(Wrap.classMemberWrap(compileSource));

        default
            -> state.outerMap.wrapInTrialClass(Wrap.methodWrap(compileSource));
    };
```

(8) As above, but with the case constants one to a line:

```
    OuterWrap outer = switch (probableKind) {
        case IMPORT
            -> state.outerMap.wrapImport(Wrap.simpleWrap(compileSource), snip);

        case EXPRESSION
            -> state.outerMap.wrapInTrialClass(Wrap.methodReturnWrap(compileSource));

        case VAR,
             TYPE_DECL,
             METHOD
            -> state.outerMap.wrapInTrialClass(Wrap.classMemberWrap(compileSource));

        default
            -> state.outerMap.wrapInTrialClass(Wrap.methodWrap(compileSource));
    };
```

**Comment by briangoetz** [comment]
(5) is fine if it fits within reasonable line limits, and is what we 
should recommend as the starting point.

(6) is the natural way to unwrap (5) when the lines get long. 
(Personally I'm largely indifferent to 6/7/8; they are all acceptable, 
and I think the distinction is mostly based on how dense and nonuniform 
the consequences are.)

If the consequences are "similar" to each other, or they are simple, 
then 6 is a winner.  If they start to diverge in subtle ways, adding 
more whitespace may make it easier for readers to understand what is 
going on, which might pull us towards 7/8.

Like with the previous choices, I might be inclined to do 7 for myself, 
but I wouldn't want to require others to code that way. I think its in 
the category of "use more whitespace if you think it makes the code more 
readable."

So:

  - 5 if it fits
  - 6 if you need a little more space
  - 7 or 8 if (6) results in something too dense to read

One thing I think you've omitted, but which we should cover, is whether 
the -> goes on the line with the case, or the consequence:

     case FOO ->
         fooAction()

vs

     case FOO
        -> fooAction()

I think this is similar to advice for joining two long boolean 
expressions with &&, or chains of method invocations; it is more obvious 
that the second line is a continuation of the first when the operator is 
at the left margin rather than the right.  So I prefer the second to the 
first (as all your examples used.)

**Comment by amaembo** [comment]
@briangoetz @stuart-marks thank you for your comments. I don't like 2) because the first case is significantly longer than the other ones. As a result, there's too much of whitespace on the subsequent lines, and it's hard to follow visually from the case label to the resulting expression without the editor's help (e. g. highlighting the current line). 4) solves this problem, so for code reading, it looks the best.

There's another consideration though: the code maintenance cost. I'm not sure about other editors but AFAIK, IntelliJ IDEA doesn't help you to maintain the vertical alignment for switch cases (we have a similar option for variable declarations, probably we should support switch cases as well), so any edits of the surrounding code may require manual reformatting, and autoformatting accidentally applied to this code may screw the things up.

As for 5-8, there's also an option to align arrows vertically:

(9)
```java
OuterWrap outer = switch (probableKind) {
    case IMPORT                 -> state.outerMap.wrapImport(Wrap.simpleWrap(compileSource), snip);
    case EXPRESSION             -> state.outerMap.wrapInTrialClass(Wrap.methodReturnWrap(compileSource));
    case VAR, TYPE_DECL, METHOD -> state.outerMap.wrapInTrialClass(Wrap.classMemberWrap(compileSource));
    default                     -> state.outerMap.wrapInTrialClass(Wrap.methodWrap(compileSource));
};
````

In this case, the longest line has 114 characters, which looks within the surrounding code style.
We can also put labels one per line, line in 4. This makes lines even shorter:
(10)
```java
OuterWrap outer = switch (probableKind) {
    case IMPORT     -> state.outerMap.wrapImport(Wrap.simpleWrap(compileSource), snip);
    case EXPRESSION -> state.outerMap.wrapInTrialClass(Wrap.methodReturnWrap(compileSource));
    case VAR, 
         TYPE_DECL, 
         METHOD     -> state.outerMap.wrapInTrialClass(Wrap.classMemberWrap(compileSource));
    default         -> state.outerMap.wrapInTrialClass(Wrap.methodWrap(compileSource));
};
```

By the way, I found that there's a redundant "preview" warning suppression at line 230. If we remove it and inline the variable, there's another opportunity for an expression switch. See the updated PR.
